### PR TITLE
uplink: integrate new Metainfo calls

### PR DESCRIPTION
**Pull request by mniewrzal**
What: Client side is adjusted to use new Metainfo methods.

Why: This is part of Metainfo refactoring. 
https://storjlabs.atlassian.net/browse/V3-2060
https://storjlabs.atlassian.net/browse/V3-2061

Please describe the tests:
 - Test 1: current tests should pass without significant changes
 - Test 2: some tests were removed because make no sense with the current design
 
Please describe the performance impact: TBD

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: pkg/pb/metainfo.proto [hunk 1]
```diff
@@ -387,7 +387,6 @@ message SegmentPosition {
 
 message SegmentBeginRequest {
     bytes           stream_id = 1 [(gogoproto.customtype) = "StreamID", (gogoproto.nullable) = false];
-    SegmentPosition position = 2;
```
Comment (egonelbre) [inline pkg/pb/metainfo.proto]: I think we need to keep segment position here to avoid later problems with tracking partial segments.

File: pkg/pb/metainfo.proto [hunk 2]
```diff
@@ -450,8 +450,9 @@ message SegmentBeginDeleteRequest {
 }
 
 message SegmentBeginDeleteResponse {
-    bytes segment_id = 1 [(gogoproto.customtype) = "SegmentID", (gogoproto.nullable) = false];
-    repeated AddressedOrderLimit addressed_limits = 2;
+    bytes                           segment_id = 1 [(gogoproto.customtype) = "SegmentID", (gogoproto.nullable) = false];
+    repeated AddressedOrderLimit    addressed_limits = 2;
+    bytes                           private_key = 3 [(gogoproto.customtype) = "PiecePrivateKey", (gogoproto.nullable) = false];
```
Comment (egonelbre) [inline pkg/pb/metainfo.proto]: Why do we need the private key for deleting?

File: pkg/pb/metainfo.proto [hunk 3]
```diff
@@ -492,4 +493,9 @@ message SegmentDownloadResponse {
     repeated AddressedOrderLimit addressed_limits = 2;
     bytes                        encrypted_inline_data = 3;
     SegmentPosition              next = 4; // can be nil
+    bytes                        private_key = 5 [(gogoproto.customtype) = "PiecePrivateKey", (gogoproto.nullable) = false];
```
Comment (egonelbre) [inline pkg/pb/metainfo.proto]: Also here?

File: pkg/storj/object_list_item.go [hunk 4]
```diff
@@ -17,4 +17,7 @@ type ObjectListItem struct {
 	ExpiresAt              time.Time
 	EncryptedMetadataNonce Nonce
 	EncryptedMetadata      []byte
+
+	// TODO should this be here ??
+	IsPrefix bool
```
Comment (egonelbre) [inline pkg/storj/object_list_item.go]: Yeah, sounds reasonable.

File: uplink/eestream/encode.go [hunk 5]
```diff
@@ -106,6 +107,17 @@ func NewRedundancyStrategyFromProto(scheme *pb.RedundancyScheme) (RedundancyStra
 	return NewRedundancyStrategy(es, int(scheme.GetRepairThreshold()), int(scheme.GetSuccessThreshold()))
 }
 
+// NewRedundancyStrategyFromStorj creates new RedundancyStrategy from the given
+// storj.RedundancyScheme.
+func NewRedundancyStrategyFromStorj(scheme storj.RedundancyScheme) (RedundancyStrategy, error) {
```
Comment (egonelbre) [inline uplink/eestream/encode.go]: `RedundancyStrategyToProto`

File: uplink/metainfo/client.go [hunk 6]
```diff
@@ -74,8 +75,8 @@ func (client *Client) Close() error {
 	return nil
 }
 
-// CreateSegment requests the order limits for creating a new segment
-func (client *Client) CreateSegment(ctx context.Context, bucket string, path storj.Path, segmentIndex int64, redundancy *pb.RedundancyScheme, maxEncryptedSegmentSize int64, expiration time.Time) (limits []*pb.AddressedOrderLimit, rootPieceID storj.PieceID, piecePrivateKey storj.PiecePrivateKey, err error) {
+// CreateSegmentOld requests the order limits for creating a new segment
```
Comment (egonelbre) [inline uplink/metainfo/client.go]: Let's do this rename in a separate PR.

File: pkg/pb/metainfo.proto (hunk 1 above)
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: The issue is that with the last segment we know its the last one after upload.

Comment (egonelbre) [reply pkg/pb/metainfo.proto]: Not after multipart upload.

File: pkg/pb/metainfo.proto (hunk 2 above)
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: hmm you tell me:) piecestore client has it as a parameter for deleting. It was added with new order signing thing.

File: pkg/pb/metainfo.proto (hunk 3 above)
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: for downloading its used for signing orders

File: uplink/metainfo/client.go (hunk 6 above)
Comment (mniewrzal) [reply uplink/metainfo/client.go]: ok

File: pkg/pb/metainfo.proto (hunk 1 above)
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: I see, I need figure out how to handle last segment when `position` is with `SegmentBeginRequest`

File: pkg/pb/metainfo.proto (hunk 2 above)
Comment (egonelbre) [reply pkg/pb/metainfo.proto]: Ah, that's the piece private key... I was thinking it was something else. Yeah, this is mainly for consistency with all other things.

File: pkg/pb/metainfo.proto (hunk 3 above)
Comment (egonelbre) [reply pkg/pb/metainfo.proto]: Yeah, forgot that it's not sent inside the order limit.

File: uplink/metainfo/client.go (hunk 6 above)
Comment (mniewrzal) [reply uplink/metainfo/client.go]: I mean, yeah maybe

File: pkg/pb/metainfo.proto (hunk 3 above)
Comment (egonelbre) [reply pkg/pb/metainfo.proto]: Can we change the order of the fields so they make a little bit more sense?

Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: addressed_limits
private_key
encrypted_inline_data
next
??

Comment (egonelbre) [reply pkg/pb/metainfo.proto]: Yeah, something like that.

File: satellite/metainfo/metainfo_test.go [hunk 7]
```diff
@@ -1015,8 +952,8 @@ func TestBeginCommitListSegment(t *testing.T) {
 		require.NoError(t, err)
 		require.Len(t, objects, 1)
 
-		require.Equal(t, params.EncryptedPath, objects[0].EncryptedPath)
-		require.Equal(t, params.EncryptedMetadata, objects[0].EncryptedMetadata)
+		// require.Equal(t, params.EncryptedPath, objects[0].EncryptedPath)
```
Comment (ifraixedes) [inline satellite/metainfo/metainfo_test.go]: Can we delete these lines?

File: uplink/metainfo/kvmetainfo/objects_test.go [hunk 8]
```diff
@@ -391,235 +393,236 @@ func TestListObjects(t *testing.T) {
...
-			}, {
-				options: optionsRecursive("", "b/yaa", storj.Before, 5),
-				more:    true,
-				result:  []string{"a/xbb", "a/xc", "aa", "b", "b/ya"},
-			}, {
-				options: options("a/", "xbb", storj.Before, 2),
-				more:    true,
-				result:  []string{"xaa", "xb"},
 			},
+			//{
```
Comment (ifraixedes) [inline uplink/metainfo/kvmetainfo/objects_test.go]: Do we plan to restore this in a short time?

File: uplink/storage/segments/store_test.go [hunk 9]
```diff
@@ -223,56 +29,61 @@ func TestCalcNeededNodes(t *testing.T) {
...
-		rs, err := eestream.NewRedundancyStrategy(eestream.NewRSScheme(fc, 1*memory.KiB.Int()), 0, 0)
-		require.NoError(t, err)
-
-		segmentStore := segments.NewSegmentStore(metainfo, ec, rs, 4*memory.KiB.Int(), 8*memory.MiB.Int64())
-		assert.NotNil(t, segmentStore)
-
-		test(t, ctx, planet, segmentStore)
-	})
-}
+// func runTest(t *testing.T, test func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet, streamStore streams.Store)) {
```
Comment (ifraixedes) [inline uplink/storage/segments/store_test.go]: Do we plan to restore these in a short time?

File: uplink/storage/segments/store.go [hunk 10]
```diff
@@ -42,11 +41,10 @@ type ListItem struct {
...
 type Store interface {
-	Meta(ctx context.Context, path storj.Path) (meta Meta, err error)
-	Get(ctx context.Context, path storj.Path) (rr ranger.Ranger, meta Meta, err error)
-	Put(ctx context.Context, data io.Reader, expiration time.Time, segmentInfo func() (storj.Path, []byte, error)) (meta Meta, err error)
-	Delete(ctx context.Context, path storj.Path) (err error)
-	List(ctx context.Context, prefix, startAfter, endBefore storj.Path, recursive bool, limit int, metaFlags uint32) (items []ListItem, more bool, err error)
+	Get(ctx context.Context, streamID storj.StreamID, segmentIndex int32, objectRS storj.RedundancyScheme) (rr ranger.Ranger, encryption storj.SegmentEncryption, err error)
+	Put(ctx context.Context, streamID storj.StreamID, data io.Reader, expiration time.Time, segmentInfo func() (int64, storj.SegmentEncryption, error)) (meta Meta, err error)
+	Delete(ctx context.Context, streamID storj.StreamID, segmentIndex int32) (err error)
+	// List(ctx context.Context, streamID storj.StreamID, prefix, startAfter, endBefore storj.Path, recursive bool, limit int, metaFlags uint32) (items []ListItem, more bool, err error)
```
Comment (ifraixedes) [inline uplink/storage/segments/store.go]: Can we get rid of it?

File: uplink/storage/segments/store.go [hunk 11]
```diff
@@ -68,133 +66,103 @@ func NewSegmentStore(metainfo *metainfo.Client, ec ecclient.Client, rs eestream.
...
 
-	bucket, objectPath, segmentIndex, err := splitPathFragments(path)
+	segmentIndex, encryption, err := segmentInfo()
 	if err != nil {
-		return Meta{}, err
+		return Meta{}, Error.Wrap(err)
 	}
 
-	savedPointer, err := s.metainfo.CommitSegment(ctx, bucket, objectPath, segmentIndex, pointer, originalLimits)
+	uploadResults := make([]*pb.SegmentPieceUploadResult, 0)
```
Comment (ifraixedes) [inline uplink/storage/segments/store.go]: let's initialize the slice with the capacity `make([]*pb.SegmentPieceUploadResult, 0, len(successfulNode))`

File: uplink/storage/segments/store.go [hunk 12]
```diff
@@ -211,179 +179,87 @@ func (s *segmentStore) Get(ctx context.Context, path storj.Path) (rr ranger.Rang
...
-		num, err := strconv.Atoi(segmentComp[1:])
-		if err != nil {
-			return -2, Error.Wrap(err)
-		}
-		return int64(num), nil
-	default:
-		return -2, Error.New("invalid segment component: %s", segmentComp)
-	}
-}
+// func convertMeta(pr *pb.Pointer) Meta {
```
Comment (ifraixedes) [inline uplink/storage/segments/store.go]: Can we get rid of these lines?

File: uplink/storage/streams/store.go [hunk 13]
```diff
@@ -462,11 +463,13 @@ func pathForKey(raw string) paths.Unencrypted {
...
 	defer mon.Task()(&ctx)(&err)
 
-	if metaFlags&meta.Size != 0 {
-		// Calculating the stream's size require also the user-defined metadata,
-		// where stream store keeps info about the number of segments and their size.
-		metaFlags |= meta.UserDefined
-	}
+	// TODO use flags with listing
+
+	// if metaFlags&meta.Size != 0 {
```
Comment (ifraixedes) [inline uplink/storage/streams/store.go]: Is this code a companion of the previous TODO?

File: uplink/storage/streams/store.go [hunk 14]
```diff
@@ -636,25 +635,13 @@ func decryptRanger(ctx context.Context, rr ranger.Ranger, decryptedSize int64, c
...
+		err := s.segments.Delete(ctx, streamID, int32(i))
 		if err != nil {
-			zap.S().Warnf("Failed deleting segment %d: %v", i, err)
-			continue
-		}
-
-		err = s.segments.Delete(ctx, currentPath)
-		if err != nil {
-			zap.S().Warnf("Failed deleting segment %v: %v", currentPath, err)
+			zap.S().Warnf("Failed deleting segment %v(%d): %v", path, i, err)
```
Comment (ifraixedes) [inline uplink/storage/streams/store.go]: Why not using the structured logger?

File: uplink/storage/streams/store_test.go [hunk 15]
```diff
@@ -1,395 +1,226 @@
...
 }
 
-func TestStreamStoreDelete(t *testing.T) {
-	ctrl := gomock.NewController(t)
-	defer ctrl.Finish()
+func runTest(t *testing.T, test func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet, streamsStore streams.Store)) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount: 1, StorageNodeCount: 4, UplinkCount: 1,
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		// TODO move apikey creation to testplanet
```
Comment (ifraixedes) [inline uplink/storage/streams/store_test.go]: out of curiosity, why this?

File: pkg/pb/metainfo.proto [hunk 16]
```diff
@@ -297,9 +297,12 @@ message ObjectListRequest {
     bytes     bucket = 1;
     bytes     encrypted_prefix = 2;
     bytes     encrypted_cursor = 3;
-    int32     limit = 4;
+    bool      recursive = 4;
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: I was not used before this PR so no backward incompatibility

File: pkg/pb/metainfo.proto [hunk 17]
```diff
@@ -352,9 +355,6 @@ message SatStreamID {
     google.protobuf.Timestamp creation_date = 5  [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
     google.protobuf.Timestamp expiration_date = 6  [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
 
-    bytes  encrypted_metadata_nonce = 7 [(gogoproto.customtype) = "Nonce", (gogoproto.nullable) = false];
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: I was not used before this PR so no backward incompatibility

File: pkg/storj/segment.go [hunk 18]
```diff
@@ -17,6 +17,16 @@ type SegmentListItem struct {
...
 	SegmentID           SegmentID
+	Size                int64
 	EncryptedInlineData []byte
 	Next                SegmentPosition
+	PiecePrivateKey     PiecePrivateKey
+
+	SegmentEncryption SegmentEncryption
+}
+
+// SegmentEncryption TODO
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply pkg/storj/segment.go]: I missed that, its about adding description

File: satellite/metainfo/metainfo.go [hunk 19]
```diff
@@ -1295,18 +1306,13 @@ func (endpoint *Endpoint) CommitSegment(ctx context.Context, req *pb.SegmentComm
...
 
-	if len(segmentID.OriginalOrderLimits) < len(req.UploadResult) {
-		return nil, status.Errorf(codes.InvalidArgument, "invalid number of upload results: wanted max %d got %d",
-			len(segmentID.OriginalOrderLimits), len(req.UploadResult))
-	}
-
-	pieces := make([]*pb.RemotePiece, len(req.UploadResult))
-	for i, result := range req.UploadResult {
-		pieces[i] = &pb.RemotePiece{
+	pieces := make([]*pb.RemotePiece, 0)
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply satellite/metainfo/metainfo.go]: yes, thanks

File: uplink/storage/segments/store_test.go (hunk 9 above)
Comment (mniewrzal) [reply uplink/storage/segments/store_test.go]: tests from `uplink/storage/segments/store_test.go` were moved and adjusted to `uplink/storage/segments/store_test.go`

Comment (mniewrzal) [reply uplink/storage/segments/store_test.go]: I need to delete this

File: uplink/storage/streams/store_test.go (hunk 15 above)
Comment (mniewrzal) [reply uplink/storage/streams/store_test.go]: copy/paste from refactored test. Its about making api key creation and access easier. To add apikey you need to make a lot of steps in test, see code below.

File: uplink/storage/segments/store.go (hunk 12 above)
Comment (mniewrzal) [reply uplink/storage/segments/store.go]: removed

File: uplink/storage/streams/store.go (hunk 13 above)
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: this is something what most probably will be fixed in next PRs. TODO is to rest of method body below.

File: uplink/storage/streams/store.go (hunk 14 above)
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: done

File: uplink/storage/streams/store.go (hunk 13 above)
Comment (ifraixedes) [reply uplink/storage/streams/store.go]: I was not fully clear because there is a blank line between the 2

Comment (mniewrzal) [reply uplink/storage/streams/store.go]: ok, I will change that

File: pkg/pb/metainfo.proto (hunk 1 above)
Comment (mniewrzal) [reply pkg/pb/metainfo.proto]: @egonelbre at the moment I have only one idea to send each segment with normal indexes (from 0 to n) and with `CommitObject` rename the last one to `l/`.

File: uplink/storage/streams/store.go [hunk 20]
```diff
@@ -641,7 +641,7 @@ func (s *streamStore) cancelHandler(ctx context.Context, streamID storj.StreamID
 	for i := int64(0); i < totalSegments; i++ {
 		err := s.segments.Delete(ctx, streamID, int32(i))
 		if err != nil {
-			zap.S().Warnf("Failed deleting segment %v(%d): %v", path, i, err)
+			zap.S().Warn("Failed deleting segment", zap.String("path", path.String()), zap.Int64("segmentIndex", i), zap.Error(err))
```
Comment (ifraixedes) [inline uplink/storage/streams/store.go]: there isn't a need for using the sugared, so let's changed it to `zap.L()`

File: satellite/metainfo/metainfo.go [hunk 21]
```diff
@@ -1670,7 +1670,7 @@ func (endpoint *Endpoint) DownloadSegment(ctx context.Context, req *pb.SegmentDo
 		if segmentMeta != nil {
 			encryptedKeyNonce, err = storj.NonceFromBytes(segmentMeta.KeyNonce)
 			if err != nil {
-				return nil, status.Errorf(codes.Internal, err.Error())
+				return nil, status.Errorf(codes.Internal, "unable to get encryption key nonce from metadata: %v", err.Error())
```
Comment (ifraixedes) [inline satellite/metainfo/metainfo.go]: We are discussing in slack about to log the error for this specific internal errors https://storj.slack.com/archives/CAQV0AE2Z/p1564567179385100

File: uplink/storage/streams/store.go (hunk 20 above)
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: as yes, good point

File: uplink/metainfo/kvmetainfo/objects_test.go (hunk 8 above)
Comment (mniewrzal) [reply uplink/metainfo/kvmetainfo/objects_test.go]: no, we are not supporting `direction` with object listing at the moment

Comment (mniewrzal) [reply uplink/metainfo/kvmetainfo/objects_test.go]: I will delete it

File: uplink/metainfo/kvmetainfo/objects_test.go [hunk 22]
```diff
@@ -391,234 +391,6 @@ func TestListObjects(t *testing.T) {
...
-				options: options("a/", "xb", storj.Before, 0),
-				result:  []string{"xa", "xaa"},
-			}, {
-				options: optionsRecursive("", "b/yaa", storj.Before, 5),
-				more:    true,
-				result:  []string{"a/xbb", "a/xc", "aa", "b", "b/ya"},
-			}, {
-				options: options("a/", "xbb", storj.Before, 2),
-				more:    true,
-				result:  []string{"xaa", "xb"},
```
Comment (kaloyan-raev) [inline uplink/metainfo/kvmetainfo/objects_test.go]: Why are these test cases removed?

Comment (mniewrzal) [reply uplink/metainfo/kvmetainfo/objects_test.go]: At the moment we don't support `direction` for objects listing with new metainfo requests.

Comment (kaloyan-raev) [reply uplink/metainfo/kvmetainfo/objects_test.go]: Will the next PRs add such support? In such, wouldn't it be better to comment these tests instead of removing them?

Comment (mniewrzal) [reply uplink/metainfo/kvmetainfo/objects_test.go]: There are no plans for such PR but I'm ok with leaving this tests commented until we will be sure about `direction` future.

File: uplink/storage/streams/store.go [hunk 23]
```diff
@@ -104,52 +104,62 @@ func (s *streamStore) Put(ctx context.Context, path Path, pathCipher storj.Ciphe
...
-		return Meta{}, currentSegment, err
+		return Meta{}, currentSegment, streamID, err
 	}
 	encPath, err := encryption.EncryptPath(path.Bucket(), path.UnencryptedPath(), pathCipher, s.encStore)
 	if err != nil {
-		return Meta{}, currentSegment, err
+		return Meta{}, currentSegment, streamID, err
+	}
+
+	streamID, err = s.metainfo.BeginObject(ctx, metainfo.BeginObjectParams{
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: I'm not sure I understand, `streamID` is return var and its used later during upload and in `defer` to handle cancelation.

File: uplink/storage/streams/store.go [hunk 24]
```diff
@@ -104,52 +104,62 @@ func (s *streamStore) Put(ctx context.Context, path Path, pathCipher storj.Ciphe
...
+		Bucket:        []byte(path.Bucket()),
+		EncryptedPath: []byte(encPath.Raw()),
+		ExpiresAt:     expiration,
+	})
+	if err != nil {
+		return Meta{}, currentSegment, streamID, err
 	}
 
+	defer func() {
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: If we didn't reach a place where we get `streamID` then make no sense to handle cancelation and trigger `defer`

File: uplink/storage/streams/store.go [hunk 25]
```diff
@@ -505,29 +500,26 @@ func (s *streamStore) List(ctx context.Context, prefix Path, startAfter, endBefo
...
-	items = make([]ListItem, len(segments))
-	for i, item := range segments {
+	items = make([]ListItem, len(objects))
+	for i, item := range objects {
 		var path Path
 		var itemPath string
 
 		if needsEncryption {
-			itemPath, err = encryption.DecryptPathRaw(item.Path, pathCipher, prefixKey)
+			itemPath, err = encryption.DecryptPathRaw(string(item.EncryptedPath), pathCipher, prefixKey)
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: `item.EncryptedPath` is just a `[]byte`

File: uplink/storage/segments/store_test.go [hunk 26]
```diff
@@ -4,212 +4,18 @@
...
-					assert.True(t, meta.Modified.After(beforeModified))
-				} else {
-					require.Contains(t, err.Error(), test.err)
-				}
-			})
-		})
-	}
-}
-
-func TestSegmentStorePutGet(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/segments/store_test.go]: Those tests were moved to `streams/store_test.go` and adjusted to new metainfo API. In general with new metainfo API it make no sense to test put/get on segment level because now we need also to create object to be able to create segment.

File: uplink/metainfo/kvmetainfo/objects_test.go [hunk 27]
```diff
@@ -278,9 +278,9 @@ func TestListObjectsEmpty(t *testing.T) {
 		assert.EqualError(t, err, "kvmetainfo: invalid direction 0")
 
 		for _, direction := range []storj.ListDirection{
-			storj.Before,
-			storj.Backward,
-			storj.Forward,
+			// storj.Before,
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/metainfo/kvmetainfo/objects_test.go]: done

File: uplink/storage/streams/store_test.go [hunk 28]
```diff
@@ -0,0 +1,226 @@
...
+		require.False(t, more)
+		require.Equal(t, len(objects), len(items))
+
+		// should list first two and more = true
+		items, more, err = streamStore.List(ctx, prefix, "", "", storj.EncNull, true, 2, meta.None)
+		require.NoError(t, err)
+		require.True(t, more)
+		require.Equal(t, 2, len(items))
+
+		// // should list only prefixes
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/streams/store_test.go]: fixed

File: uplink/storage/streams/store_test.go [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+		path       string
+		metadata   []byte
+		expiration time.Time
+		content    []byte
+	}{
+		{"test inline put/get", "path/1", []byte("inline-metadata"), time.Time{}, testrand.Bytes(2 * memory.KiB)},
+		{"test remote put/get", "mypath/1", []byte("remote-metadata"), time.Time{}, testrand.Bytes(100 * memory.KiB)},
+	} {
+		test := tt
+		runTest(t, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet, streamStore streams.Store) {
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/streams/store_test.go]: good idea, done

File: uplink/storage/streams/store_test.go [hunk 30]
```diff
@@ -0,0 +1,226 @@
...
+		path       string
+		metadata   []byte
+		expiration time.Time
+		content    []byte
+	}{
+		{"test inline delete", "path/1", []byte("inline-metadata"), time.Time{}, testrand.Bytes(2 * memory.KiB)},
+		{"test remote delete", "mypath/1", []byte("remote-metadata"), time.Time{}, testrand.Bytes(100 * memory.KiB)},
+	} {
+		test := tt
+		runTest(t, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet, streamStore streams.Store) {
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply uplink/storage/streams/store_test.go]: good idea, done

File: uplink/storage/segments/store.go [hunk 31]
```diff
@@ -218,179 +184,77 @@ func (s *segmentStore) Get(ctx context.Context, path storj.Path) (rr ranger.Rang
...
 // CalcNeededNodes calculate how many minimum nodes are needed for download,
 // based on t = k + (n-o)k/o
-func CalcNeededNodes(rs *pb.RedundancyScheme) int32 {
+func CalcNeededNodes(rs storj.RedundancyScheme) int32 {
 	extra := int32(1)
 
-	if rs.GetSuccessThreshold() > 0 {
-		extra = ((rs.GetTotal() - rs.GetSuccessThreshold()) * rs.GetMinReq()) / rs.GetSuccessThreshold()
+	if rs.OptimalShares > 0 {
+		extra = int32(((rs.TotalShares - rs.OptimalShares) * rs.RequiredShares) / rs.OptimalShares)
```
Comment (BlackDuck888) [inline uplink/storage/segments/store.go]: https://storjlabs.atlassian.net/browse/V3-2485
For download, I am convinced that TotalShares should be removed from the formula. TotalShares are theoretical but never existing shares, just needed to query enough nodes for upload.

Comment (mniewrzal) [reply uplink/storage/segments/store.go]: I would prefer to make such change as a separate PR

Comment (BlackDuck888) [reply uplink/storage/segments/store.go]: However, I saw that you touched that and only want point to V3-2485.

File: uplink/metainfo/kvmetainfo/stream.go [hunk 32]
```diff
@@ -19,6 +17,7 @@ var _ storj.ReadOnlyStream = (*readonlyStream)(nil)
 type readonlyStream struct {
 	db *DB
 
+	ID        storj.StreamID
```
Comment (egonelbre) [inline uplink/metainfo/kvmetainfo/stream.go]: does this have to be public?

Comment (mniewrzal) [reply uplink/metainfo/kvmetainfo/stream.go]: no

File: uplink/storage/segments/store.go [hunk 33]
```diff
@@ -72,133 +69,102 @@ func NewSegmentStore(metainfo *metainfo.Client, ec ecclient.Client, rs eestream.
...
+		return Meta{}, Error.Wrap(err)
+	}
 
-		originalLimits = make([]*pb.OrderLimit, len(limits))
-		for i, addressedLimit := range limits {
-			originalLimits[i] = addressedLimit.GetLimit()
-		}
+	segmentID, limits, piecePrivateKey, err := s.metainfo.BeginSegment(ctx, metainfo.BeginSegmentParams{
+		StreamID:     streamID,
+		MaxOderLimit: s.maxEncryptedSegmentSize,
```
Comment (egonelbre) [inline uplink/storage/segments/store.go]: typo in `MaxOderLimit`

File: uplink/storage/streams/store.go [hunk 34]
```diff
@@ -418,42 +413,38 @@ func (s *streamStore) Delete(ctx context.Context, path Path, pathCipher storj.Ci
...
+		Bucket:        []byte(path.Bucket()),
+		EncryptedPath: []byte(encPath.Raw()),
+	})
 	if err != nil {
 		return err
 	}
 
-	streamInfo, streamMeta, err := TypedDecryptStreamInfo(ctx, lastSegmentMeta.Data, path, s.encStore)
+	// TODO handle `more`
+	items, _, err := s.metainfo.ListSegmentsNew(ctx, metainfo.ListSegmentsParams{
```
Comment (egonelbre) [inline uplink/storage/streams/store.go]: I think we need to handle this, otherwise we'll start leaking data after 6.4GB sized uploads, but I'm fine with doing a separate PR for this.

File: uplink/storage/streams/store.go [hunk 35]
```diff
@@ -418,42 +413,38 @@ func (s *streamStore) Delete(ctx context.Context, path Path, pathCipher storj.Ci
 		return err
 	}
 
-	lastSegmentPath, err := createSegmentPath(ctx, -1, path.Bucket(), encPath)
-	if err != nil {
-		return err
-	}
+	// TODO maybe for now just use GetObject and decrypt num of segments from metainfo
+	// to avoid two satellite requestes ??
```
Comment (egonelbre) [inline uplink/storage/streams/store.go]: The reason we don't want to ask a "number of segments" is because with multipart uploads we won't know the part numbers, so we would end up needing to do the listing anyways. But we could batch begin delete and list segments together into a single request.

File: uplink/storage/streams/store.go (hunk 34 above)
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: I will prepare PR for that when this will be merged.

File: uplink/storage/streams/store.go (hunk 35 above)
Comment (mniewrzal) [reply uplink/storage/streams/store.go]: yeah, this was only for the current implementation. For batching we still need to implement passing params between requests/response but yes this can be done.

**Comment by egonelbre** [comment]
- [x] Does the PR describe what changes are being made?
- [x] Does the PR describe why the changes are being made?
- [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
- [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
- [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
- [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
- [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
- [x] Does any documentation need updating?
- [x] Do the database access patterns make sense?
